tortoisebeacon/tortoise_beacon.go

```diff
@@ -503,84 +503,61 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
 	// For next rounds,
 	// wait for δ time, and construct a message that points to all messages from previous round received by δ.
 	// rounds 1 to K
-	timer := time.NewTimer(tb.config.FirstVotingRoundDuration + tb.config.WeakCoinRoundDuration)
+	timer := time.NewTimer(tb.config.FirstVotingRoundDuration)
 	defer timer.Stop()
 
 	var (
-		coinFlip            bool
-		ownLastRoundVotesMu sync.RWMutex
-		ownLastRoundVotes   allVotes
+		ownVotes allVotes
+		err      error
 	)
-
 	for round := firstRound; round <= tb.lastRound(); round++ {
-		// always use coinflip from the previous round for current round.
-		// round 1 is running without coinflip (e.g. value is false) intentionally
 		round := round
-		previousCoinFlip := coinFlip
+		votes := ownVotes
 		tb.eg.Go(func() error {
 			if round == firstRound {
 				if err := tb.sendProposalVote(ctx, epoch); err != nil {
-					logger.With().Error("Failed to send proposal vote",
+					logger.With().Error("failed to send proposal vote",
+						log.Uint32("round_id", uint32(round)),
+						log.Err(err))
+				}
+			} else {
+				if err := tb.sendFollowingVote(ctx, epoch, round, votes); err != nil {
+					logger.With().Error("failed to send following vote",
 						log.Uint32("round_id", uint32(round)),
 						log.Err(err))
-
-					return fmt.Errorf("failed to send proposal vote: %w", err)
 				}
-
-				return nil
-			}
-
-			// next rounds, send vote
-			// construct a message that points to all messages from previous round received by δ
-			ownCurrentRoundVotes, err := tb.calcVotes(epoch, round, previousCoinFlip)
-			if err != nil {
-				logger.With().Error("Failed to calculate votes",
-					log.Uint32("round_id", uint32(round)),
-					log.Err(err))
-
-				return fmt.Errorf("calculate votes: %w", err)
-			}
-
-			if round == tb.lastRound() {
-				ownLastRoundVotesMu.Lock()
-				ownLastRoundVotes = ownCurrentRoundVotes
-				ownLastRoundVotesMu.Unlock()
-			}
-
-			if err := tb.sendFollowingVote(ctx, epoch, round, ownCurrentRoundVotes); err != nil {
-				logger.With().Error("Failed to send following vote",
-					log.Uint32("round_id", uint32(round)),
-					log.Err(err))
-
-				return fmt.Errorf("send following vote: %w", err)
 			}
-
-			return nil
-		})
-
-		tb.eg.Go(func() error {
-			tb.startWeakCoinRound(ctx, epoch, round)
 			return nil
 		})
+		if round != firstRound {
+			tb.eg.Go(func() error {
```

Comment:
i think there is still a potential race condition here. a malicious peer can still wait till the weakcoin round to finish before it send out the votes, and their late votes will still be counted.

i think we should
- send vote
- wait for the duration for vote to come in
- start weak coin (this is the cutoff time for counting votes in this round)
- count votes without weak coin
- finish weak coin
- count undecided votes with weak coin
- the late vote in this round will be added to the next round

Reply:
that makes sense

Reply:
@countvonzero i updated the code. why do you think that we should:
> the late vote received after weak coin started will be added to the next round
why not to ignore that vote? in the next round we will have a next set of votes

Reply:
that was basedd on Tal in the R&D meeting. my impression is tgat we want as many vote as possible to cross threshold. as long as a node isnt double voting, we should count it. but if that introduce extra code complexity, we should talk to Tal abt this

Reply:
i think thats possible in general but we need to revisit how votes are buffered. but what i don't understand is what we should do with the vote that will be sent in the next round in such case? i don't think that we should count votes from the same smesher twice (e.g. first time from late vote in the previous round and second time from timely vote in the current round)

Reply:
it's not so much `count votes from the same smesher twice` tho. essentially every smesher gets to vote K times where K is the number of rounds. 

if we consider a honest smesher (not casting multiple votes for the same round), then it's just network delay and there is not reason to _not_ count it.

but if a smesher is malicious, casting late votes (after it learns the weakcoin) can potentially sway the vote to cross the threshold early in that round.

my guess would be it's OK to just drop late votes for code simplicity. but it's worth to double-check with Tal maybe in tuesday R&D meeting

Reply:
sure lets discuss it. i can also open a follow up issue to refactor buffering to make counting late nodes easier